### PR TITLE
Extend pykostal to support multiple dxsIds and restructred response

### PR DESCRIPTION
**Pull request by bedmanforever**
- Added unit to const data structure
- Use const data structure as return data structure
- Support for multiple entry queries
- Extended the request parameter generation to allow querying of multiple dxsIDs
- Added result query generation to return only queried values
ToDo:
- Adding work around to overcome the 25 entries limit of the Kostal API

### CODE REVIEW

**Comment by bedmanforever** [comment]
Fixed 25 dxsIds limit

**Comment by scheidtdav** [comment]
Hey,

thanks for the PR and sorry for the long delay. I will have a deeper look into this later today.
In any case, I think this needs to go back to the [original repository](https://github.com/DAMEK86/pykostal), such that @DAMEK86 can release a new version :)

**Comment by bedmanforever** [comment]
Hey,

thanks for the review. The challenge is that the changes are breaking the
outputs of the API Calls, therefore we should align if the changes are
valid from our needs, before we pull it into the original repository.

David Scheidt ***@***.***> schrieb am Do., 11. Nov. 2021,
11:52:

File: kostal/__init__.py [hunk 1]
```diff
@@ -97,27 +104,219 @@ def __init__(self, session: aiohttp.ClientSession, url: str = None, username: st
...
 
-    async def __fetch_dxs_entry(self, entry_id: []):
-        r = await self.__fetch_data({'dxsEntries': entry_id})
-        return DxsResponse(**r).get_entry_by_id(entry_id)
+    async def __fetch_dxs_entry(self, entry_ids: []):
+        if len(entry_ids) > 0:
+            # Kostal API is only returning max 25 entries - if entry_ids contains more than 25 entries we will split up the request
+            splits = round(len(entry_ids) / 25) + 1
+            dxsEntries_result = {"dxsEntries": [], "session": {}, "status": {}}
+            for i in range(0, splits):
```
Comment (scheidtdav) [inline kostal/__init__.py:130]: If you like you may be able to simplify this even further:
1. Split the entry_ids into chunks of 25 elements: `splits = [ entry_ids[pos:pos+25 for pos in range(0, len(entry_ids), 25] ]`
2. Generate the request per chunk: ` requests = [ { "dxsEntries": [id for id in split], "session": {}, "status": {} } for split in splits ] `
3. Iterate the prepared requests:
```
for request in requests:
    result = await self.__fetch_data(request)
    dxsEntries_result[...] etc etc
```

This is however already optimization, the code works fine the way it is :)

**Comment by scheidtdav** [comment]
Valid point, will have a deeper look now!

**Review by scheidtdav — COMMENTED** [review]
Great idea to include units for the individual data points.
It also seems more use friendly to me, to be able to query data using the human-readable names instead of the dxsIds :)

I think this should work out fine for the homeassistant integration :rocket: 
Do you want to do the optimization or should we merge right away?

**Comment by DAMEK86** [comment]
I'm not the pep8 guru but is it really necessary to requote the strings?
see https://www.python.org/dev/peps/pep-0008/#string-quotes

**Comment by DAMEK86** [comment]
I love your hard work :+1: 
I'm happy to bump the MINOR version of the upstream if you rework the method names ;)

**Comment by bedmanforever** [comment]
> I'm happy to bump the MINOR version of the upstream if you rework the method names ;)

Thanks for the feedback. As I am new to Open Source contributions, can you let me know a little bit more specific what you would like to have changes (method names and string quotes wise)
It's my first contribution ;-)

File: kostal/__init__.py (hunk 1 above)
Comment (bedmanforever) [reply kostal/__init__.py]: Optimization recommendations have been implemented. requests needs to be a list of tuples due to the fact that "dxsEntries" is not a unique key in the request URL
@DAMEK86 , @scheidtdav Please have a look on the proposed changes

**Comment by bedmanforever** [comment]
Commented all debugging prints to be clean for the merge

**Comment by scheidtdav** [comment]
@bedmanforever Thanks for the changes, I am good with this.

Lets continue with this in the component and see if more changes are required.
Once done, we can file a merge request for @DAMEK86 , I'd say.

**Pull request merged**